crypto/threads_pthread.c

```diff
@@ -42,12 +45,562 @@
+{
+    uint64_t ret;
+
+    pthread_mutex_lock(&atomic_sim_lock);
+    *p += v;
+    ret = *p;
+    pthread_mutex_unlock(&atomic_sim_lock);
+    return ret;
+}
+
+# define ATOMIC_ADD_FETCH(p, v, o) fallback_atomic_add_fetch(p, v)
+
+static inline uint64_t fallback_atomic_fetch_add(uint64_t *p, uint64_t v)
+{
+    uint64_t ret;
+
+    pthread_mutex_lock(&atomic_sim_lock);
+    ret = *p;
+    *p += v;
+    pthread_mutex_unlock(&atomic_sim_lock);
+    return ret;
+}
+
+# define ATOMIC_FETCH_ADD(p, v, o) fallback_atomic_fetch_add(p, v)
+
+static inline uint64_t fallback_atomic_sub_fetch(uint64_t *p, uint64_t v)
+{
+    uint64_t ret;
+
+    pthread_mutex_lock(&atomic_sim_lock);
+    *p -= v;
+    ret = *p;
+    pthread_mutex_unlock(&atomic_sim_lock);
+    return ret;
+}
+
+# define ATOMIC_SUB_FETCH(p, v, o) fallback_atomic_sub_fetch(p, v)
+
+static inline uint64_t fallback_atomic_and_fetch(uint64_t *p, uint64_t m)
+{
+    uint64_t ret;
+
+    pthread_mutex_lock(&atomic_sim_lock);
+    *p &= m;
+    ret = *p;
+    pthread_mutex_unlock(&atomic_sim_lock);
+    return ret;
+}
+
+# define ATOMIC_AND_FETCH(p, v, o) fallback_atomic_and_fetch(p, v)
+
+static inline uint64_t fallback_atomic_or_fetch(uint64_t *p, uint64_t m)
+{
+    uint64_t ret;
+
+    pthread_mutex_lock(&atomic_sim_lock);
+    *p |= m;
+    ret = *p;
+    pthread_mutex_unlock(&atomic_sim_lock);
+    return ret;
+}
+
+# define ATOMIC_OR_FETCH(p, v, o) fallback_atomic_or_fetch(p, v)
+#endif
+
+static CRYPTO_THREAD_LOCAL rcu_thr_key;
+
+/*
+ * users is broken up into 2 parts
+ * bits 0-15 current readers
+ * bit 32-63 - ID
+ */
+# define READER_SHIFT 0
+# define ID_SHIFT 32
+# define READER_SIZE 16
+# define ID_SIZE 32
+
+# define READER_MASK     (((uint64_t)1 << READER_SIZE)-1)
+# define ID_MASK         (((uint64_t)1 << ID_SIZE)-1)
+# define READER_COUNT(x) (((uint64_t)(x) >> READER_SHIFT) & READER_MASK)
+# define ID_VAL(x)       (((uint64_t)(x) >> ID_SHIFT) & ID_MASK)
+# define VAL_READER      ((uint64_t)1 << READER_SHIFT)
+# define VAL_ID(x)       ((uint64_t)x << ID_SHIFT)
+
+/*
+ * This is the core of an rcu lock. It tracks the readers and writers for the
+ * current quiescence point for a given lock. Users is the 64 bit value that
+ * stores the READERS/ID as defined above
+ *
+ */
+struct rcu_qp {
+    uint64_t users;
+};
+
+struct thread_qp {
+    struct rcu_qp *qp;
+    CRYPTO_RCU_LOCK lock;
+};
+
+#define MAX_QPS 10
+/*
+ * This is the per thread tracking data
+ * that is assigned to each thread participating
+ * in an rcu qp
+ *
+ * qp points to the qp that it last acquired
+ *
+ */
+struct rcu_thr_data {
+    struct thread_qp thread_qps[MAX_QPS];
+};
+
+/*
+ * This is the internal version of a CRYPTO_RCU_LOCK
+ * it is cast from CRYPTO_RCU_LOCK
+ */
+struct rcu_lock_st {
+    /* Callbacks to call for next ossl_synchronize_rcu */
+    struct rcu_cb_item *cb_items;
+
+    /* rcu generation counter for in-order retirement */
+    uint64_t id_ctr;
```

Comment:
Shouldn't this be `uint32_t` like `next_to_retire` to match the number of bits in `ID_SIZE`